tests/test_core/test_component.py

```diff
@@ -0,0 +1,76 @@
+import functools
+
+from winter.core import Component
+from winter.core import is_component
+from winter.core.application import WinterApplication
+from winter.core.state_key import StateKey
+
+
+def test_is_component():
+    winter_app = WinterApplication()
+
+    @winter_app.component
+    class SimpleController:
+        pass
+
+    assert is_component(SimpleController)
+
+
+def test_methods():
+    winter_app = WinterApplication()
+
+    @winter_app.component
+    class SimpleComponent:
+
+        @winter_app.component_method
+        def simple_method(self):
+            return None
+
+    controller = Component(SimpleComponent)
+
+    assert len(controller.methods) == 1, SimpleComponent.simple_method
+    method = controller.methods[0]
+
+    assert method is SimpleComponent.simple_method
+
+
+def test_method_state():
+    winter_app = WinterApplication()
+
+    def route(param: str):
+        state_key = StateKey('path')
+        state_value = param
+        return functools.partial(winter_app.component_method, state_key=state_key, state_value=state_value)
+
+    @winter_app.component
+    class SimpleComponent:
+
+        @route('/url/')
+        def simple_method(self):
+            return None
+
+    state_key = StateKey('path')
+    assert SimpleComponent.simple_method.get_state(state_key) == '/url/'
+
+
+def test_method_state_many():
+
+    winter_app = WinterApplication()
+
+    def query_param(param: str):
+        state_key = StateKey('query', many=True)
+        state_value = param
+        return functools.partial(winter_app.component_method, state_key=state_key, state_value=state_value)
```

Comment:
It's not obvious for me that `component_method` is used both to mark the method and to update it's state/metadata.